eth2/fork_choice/src/optimised_lmd_ghost.rs

```diff
@@ -0,0 +1,446 @@
+const FORK_CHOICE_BALANCE_INCREMENT: u64 = 1e9 as u64;
+const MAX_DEPOSIT_AMOUNT: u64 = 32e9 as u64;
+const EPOCH_LENGTH: u64 = 64;
+
+/// The optimised LMD-GHOST fork choice rule.
+/// NOTE: This uses u32 to represent difference between block heights. Thus this is only
+/// applicable for block height differences in the range of a u32.
+/// This can potentially be parallelized in some parts.
+// we use fast log2, a log2 lookup table is implemented in Vitaliks code, potentially do
+// the comparison. Log2_raw takes 2ns according to the documentation.
+#[inline]
+fn log2_int(x: u32) -> u32 {
+    log2_raw(x as f32) as u32
+}
+
+fn power_of_2_below(x: u32) -> u32 {
+    2u32.pow(log2_int(x))
+}
+
+/// Stores the necessary data structures to run the optimised lmd ghost algorithm.
+pub struct OptimisedLMDGhost<T: ClientDB + Sized> {
+    /// A cache of known ancestors at given heights for a specific block.
+    //TODO: Consider FnvHashMap
+    cache: HashMap<CacheKey<u32>, Hash256>,
+    /// Log lookup table for blocks to their ancestors.
+    //TODO: Verify we only want/need a size 16 log lookup
+    ancestors: Vec<HashMap<Hash256, Hash256>>,
+    /// Stores the children for any given parent.
+    children: HashMap<Hash256, Vec<Hash256>>,
+    /// The latest attestation targets as a map of validator index to block hash.
+    //TODO: Could this be a fixed size vec
+    latest_attestation_targets: HashMap<u64, Hash256>,
+    /// Block storage access.
+    block_store: Arc<BeaconBlockStore<T>>,
+    /// State storage access.
+    state_store: Arc<BeaconStateStore<T>>,
+    max_known_height: Height,
+}
+
+impl<T> OptimisedLMDGhost<T>
+where
+    T: ClientDB + Sized,
+{
+    pub fn new(
+        block_store: Arc<BeaconBlockStore<T>>,
+        state_store: Arc<BeaconStateStore<T>>,
+    ) -> Self {
+        OptimisedLMDGhost {
+            cache: HashMap::new(),
+            ancestors: vec![HashMap::new(); 16],
+            latest_attestation_targets: HashMap::new(),
+            children: HashMap::new(),
+            max_known_height: Height::new(0),
+            block_store,
+            state_store,
+        }
+    }
+
+    /// Finds the latest votes weighted by validator balance. Returns a hashmap of block_hash to
+    /// weighted votes.
+    pub fn get_latest_votes(
+        &self,
+        state_root: &Hash256,
+        block_slot: Slot,
+    ) -> Result<HashMap<Hash256, u64>, ForkChoiceError> {
+        // get latest votes
+        // Note: Votes are weighted by min(balance, MAX_DEPOSIT_AMOUNT) //
+        // FORK_CHOICE_BALANCE_INCREMENT
+        // build a hashmap of block_hash to weighted votes
+        let mut latest_votes: HashMap<Hash256, u64> = HashMap::new();
+        // gets the current weighted votes
+        let current_state = self
+            .state_store
+            .get_reader(&state_root)?
+            .ok_or_else(|| ForkChoiceError::MissingBeaconState(*state_root))?
+            .into_beacon_state()
```

Comment:
There's a short-hand method for this, just FYI

https://github.com/sigp/lighthouse/blob/e4f6fe047da5eab3303dae826b6b4dcd86dafcb6/beacon_node/db/src/stores/beacon_state_store.rs#L22

Reply:
I think it was implemented after `lmd-ghost`

Reply:
Sweet, will update